### PR TITLE
feat(rest): authenticate clients

### PR DESCRIPTION
**Pull request by paulyoong**
Provide the ability to authenticate client requests through the use of a
JSON Web Token (JWT). This requires a JSON Web Key (JWK) to be available
for use by the REST service so that it can validate the JWT.

The REST service can be started with the '--no-auth' argument to disable
authentication. This is useful for test cases where authentication is
unnecessary.

The REST service can be started with the '--jwk' argument to enable
authentication. The path to the relevant JWK file must also be provided.

Example authentication files are provided in
../Mayastor/control-plane/rest/authentication for test purposes and
should not be used in production.

Resolves: CAS-688

### CODE REVIEW

File: control-plane/rest/service/src/authentication.rs [hunk 1]
```diff
@@ -0,0 +1,122 @@
+use actix_web::{Error, HttpRequest};
+use jsonwebtoken::{crypto, Algorithm, DecodingKey};
+use std::str::FromStr;
+
+use actix_web::http::header::Header;
+use actix_web_httpauth::headers::authorization;
+use once_cell::sync::OnceCell;
+use std::fs::File;
+static JWK: OnceCell<JsonWebKey> = OnceCell::new();
```
Comment (tiagolobocastro) [inline control-plane/rest/service/src/authentication.rs:9]: Don't think we need this, you can add data to the rest server and then retrieve it on the from_request from the BearerToken.

**Review by tiagolobocastro — APPROVED** [review]
Should also add tests for invalid tokens both on the client and server?

File: control-plane/macros/actix/src/lib.rs [hunk 2]
```diff
@@ -57,6 +57,17 @@ impl Method {
...
         Ok(handler.sig.ident)
     }
+    /// Add authorisation to handler functions by adding a BearerToken as an
+    /// additional function argument.
+    /// The BearerToken is defined in
+    /// ../Mayastor/control-plane/rest/service/src/v0/mod.rs
+    fn handler_fn_with_auth(item: TokenStream) -> TokenStream2 {
+        let mut func: ItemFn = syn::parse(item).expect("Failed to parse item");
+        let new_input = syn::parse_str("_token: BearerToken")
+            .expect("Failed to parse bearer token");
```
[the remark this answers is not in the document]
Comment (paulyoong) [reply control-plane/macros/actix/src/lib.rs]: Done

File: control-plane/rest/Cargo.toml [hunk 3]
```diff
@@ -39,6 +39,9 @@ paperclip = { version = "0.5.0", default-features = false, optional = true }
 macros = { path = "../macros" }
 http = "0.2.3"
 tinytemplate = { version = "1.2" }
+jsonwebtoken = "7.2.0"
+actix-web-httpauth = "0.5.0"
+once_cell = "1.6.0"
```
[the remark this answers is not in the document]
Comment (paulyoong) [reply control-plane/rest/Cargo.toml]: The requirement for `once_cell` has now been removed now that the JWK has been added as `app_data` to the rest server.

File: control-plane/rest/service/src/authentication.rs [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+/// the sender is who they claim to be.
+pub fn authenticate(req: &HttpRequest) -> Result<(), Error> {
+    // If authentication is disabled there is nothing to do.
+    if !jwk().auth_enabled() {
+        return Ok(());
+    }
+
+    // Check the request for a bearer token.
+    // An error is returned if the bearer token cannot be found.
+    match authorization::Authorization::<authorization::Bearer>::parse(req) {
```
[the remark this answers is not in the document]
Comment (paulyoong) [reply control-plane/rest/service/src/authentication.rs]: Done - the token is now manually extracted from the request header which has removed the need for this crate.

File: control-plane/rest/service/src/authentication.rs (hunk 1 above)
Comment (paulyoong) [reply control-plane/rest/service/src/authentication.rs]: Thanks, done - this is now done and has removed the need for `once_cell`.

**Comment by paulyoong** [comment]
> Should also add tests for invalid tokens both on the client and server?

Good spot, I've added these test cases now.

**Comment by paulyoong** [comment]
bors merge

**Comment by paulyoong** [comment]
bors merge

**Comment by paulyoong** [comment]
bors merge

**Comment by tiagolobocastro** [comment]
bors merge

**Pull request merged**